wolfcrypt/src/ed25519.c

```diff
@@ -56,30 +56,47 @@ static const byte ed25519Ctx[ED25519CTX_SIZE+1] =
                                              "SigEd25519 no Ed25519 collisions";
 #endif
 
+static int ed25519_hash_reset(ed25519_key* key) {
+    int ret;
+    if (key->sha_clean_flag)
+        ret = 0;
+    else {
+        wc_Sha512Free(&key->sha);
```

Comment:
Why free before init?
Shouldn't be necessary.

Reply:
`_reset()` is only allowed on an `ed25519_key` that has been `wc_ed25519_init()`ed.  If it has been inited, then `wc_InitSha512_ex()` has been called on its sha state.  In that case, initing again without first freeing will cause a memory leak if any resources are obtained by the init.

Reply:
Better would be a wc_Sha512Reinit() that just calls the underlying Sha512Init().
That way we aren't deallocating and reallocating all the time.
Alternatively, we could always call wc_Sha512Final().

Reply:
It never deallocates-and-reallocates in normal usage, because the `.sha_clean_flag` is set by `ed25519_hash_final()`.

That is to say, absent errors, `wc_InitSha512_ex()` will be called exactly once, in `wc_ed25519_init_ex()`, and `wc_Sha512Free()` will be called exactly once, by `wc_ed25519_free()`, over the lifespan of the key in memory.  The old code called `wc_InitSha512_ex()` and `wc_Sha512Free()` for every function call.

Calling `wc_Sha512Final()` unconditionally entails frivolous reinits on the known-clean sha state.  Moreover, the incumbent implementation does not unconditionally clean the state of the sha, so cannot be used for this purpose as-is.  Indeed, in an earlier iteration of this PR, I altered the semantics of `wc_Sha512Final()` to unconditionally call `InitSha512()` for precisely this purpose.  This proved to be a poor choice and I reverted it.  See the commit log for the gory details if you're interested.

Reply:
oh, ha!  I completely forgot, earlier commits also added and removed `wc_Sha*Reset()`.  The state of the PR reflects all those lessons learned in those earlier attempts, and what we have now is the best strategy we've found so far.